pkg/compiler/check.go

```diff
@@ -577,12 +590,7 @@ func (comp *compiler) checkTypeCtors(t *ast.Type, dir prog.Dir, isArg bool,
 	ctors map[string]bool, checked map[structDir]bool) {
 	desc := comp.getTypeDesc(t)
 	if desc == typeResource {
-		// TODO(dvyukov): consider changing this to "dir == prog.DirOut".
-		// We have few questionable cases where resources can be created
-		// only by inout struct fields. These structs should be split
-		// into two different structs: one is in and second is out.
-		// But that will require attaching dir to individual fields.
-		if dir != prog.DirIn {
+		if dir == prog.DirOut {
```

Comment:
Okay, I took this change back ([here](https://github.com/google/syzkaller/pull/2008/commits/82ffbda67776502531b39b560e33ac8956d90827#diff-47a4f07e1880cd23bc05dca73d329c1dR602-R607)).

The errors rise after doing this change seem very useful to pinpoint the locations where we can utilize per field directions.

I will think more about other implications of this change, and send another PR to further discuss as you suggested.

Reply:
> The errors rise after doing this change seem very useful to pinpoint the locations where we can utilize per field directions.

That's interesting. If you can find more places to apply per-field attributes, that's definitely useful.
I suspect it may also be possible to find some places to apply the attribute by searching for [opt] resources.